virtcontainers: netmon: Monitor network changes

sboeuf opened the pull request:
This commit introduces a new watcher dedicated to the monitoring
of a specific network namespace in order to detect any change that
could happen to the network.

As a result of such a detection, the watcher should call into the
appropriate runtime path with the correct arguments to modify the
pod network accordingly.

Fixes #170

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

sboeuf commented:
This PR supersedes the original C implementation https://github.com/kata-containers/runtime/pull/194

devimc commented on virtcontainers/netmon/netmon.go:
```diff
@@ -0,0 +1,530 @@
+
+	netlinkFamily = netlink.FAMILY_V4
+
+	storageParentPath = "/var/run/kata-containers/netmon/sbs"
```
this path should be generated by in the Makefile

sboeuf replied:
Good point.

devimc commented on virtcontainers/netmon/netmon.go:
```diff
@@ -0,0 +1,530 @@
+		os.Exit(1)
+	}
+
+	/*
```
why is this commented? is this a TODO ?

sboeuf replied:
Oh yeah, I forgot to remove it. I'll do that when I'll update the PR.

sboeuf replied:
done

devimc commented on virtcontainers/netmon/netmon.go:
```diff
@@ -0,0 +1,530 @@
+		if addr.IPNet == nil {
+			continue
+		}
+	}
```
and this loop if for ... ?

sboeuf replied:
Oh yeah interesting I forgot to complete the implementation there !

sboeuf replied:
done

sboeuf commented:
Things left to do:
- Uses `logrus` package for proper logging
- Add code to virtcontainers:
  - Spawn this binary inside the netns after the netns has been created
  - Change the internal network interfaces name (tap, bridges) by adding "_kata" suffix. 
- Improve use case for crash/reboot (save/restore from filesystem)

sboeuf commented on virtcontainers/netmon/netmon.go:
```diff
@@ -0,0 +1,531 @@
+
+const (
+	netmonName    = "kata-netmon"
+	netmonVersion = "0.0.1"
```
Yes they could be generated through the build.

sboeuf replied:
done

sboeuf commented on virtcontainers/netmon/netmon.go:
```diff
@@ -0,0 +1,531 @@
+
+	kataSuffix = "kata"
+
+	netlinkFamily = netlink.FAMILY_V4
```
Yes I'll add a comment that we want to focus on ipv4 for now, just to make sure we're not introducing too much complexity at the first step.

sboeuf commented on virtcontainers/netmon/netmon.go:
```diff
@@ -0,0 +1,531 @@
+}
+
+func printVersion() {
+	fmt.Printf("%s version %s\n", netmonName, netmonVersion)
```
Sure, but this will come later, I need to get the code working and tested first.

sboeuf commented on virtcontainers/netmon/netmon.go:
```diff
@@ -0,0 +1,531 @@
+
+		src := ""
+		if netRoute.Src.To4() != nil {
+			dst = netRoute.Src.String()
```
Thanks for noticing ;)

sboeuf replied:
done

sboeuf commented on virtcontainers/netmon/netmon.go:
```diff
@@ -0,0 +1,531 @@
+
+	storageParentPath = "/var/run/kata-containers/netmon/sbs"
+	storageDirPerm    = 0750
+	sharedFile        = "shared.json"
```
done

sboeuf commented on virtcontainers/netmon/netmon.go:
```diff
@@ -0,0 +1,531 @@
+	}
+}
+
+func main() {
```
That sounds like a nice thing to do ;)

bergwolf commented:
To clarify, this depends on #287 for the kata cli network sub-command. Adding WiP because of it.

sboeuf commented:
Thanks @bergwolf ! Indeed, it relies on the fact that #287 needs to be merged first.

bergwolf commented on virtcontainers/netmon/netmon.go:
```diff
@@ -0,0 +1,531 @@
+	flag.BoolVar(&params.debug, "d", false, "enable debug mode")
+	flag.BoolVar(&version, "v", false, "display program version and exit")
+	flag.StringVar(&params.sandboxID, "s", "", "sandbox id (required)")
+	flag.StringVar(&params.runtimePath, "r", "", "runtime path (required)")
```
IMO, one benefit of calling the kata API directly is that if you modularize it properly, we can create a goroutine to monitor the netns in containerd kata shim and that presents no dependency on the kata-runtime cli, which is good from containerd kata shim perspective.

How about a structure like this:
- virtcontainers/netnsmon: a package that calls kata API to do netns monitoring
- cli/netmon: a command line built on top of virtcontainers/netnsmon. Other kata-runtime sub-command can invoke it to create a daemon to monitor the netns
- for containerd kata shim, it can create a goroutine that calls virtcontainers/netnsmon to monitor the netns when it wants to.

In both kata-runtime and containerd kata shim case, virtcontainers do not invoke netnsmon but let the callers decide and drive the procedure.

WDTY?

cc @gnawux @lifupan

ydjainopensource commented:
@sboeuf #287 has been merged now but since opentracing support is being added for kata as tracked [here](https://github.com/kata-containers/kata-containers/issues/27) , Do we wish to keep this PR open? considering its gonna be unnecessary after opentracing is added.

jodh-intel commented:
Hi @ydjainopensource - I think we'll still need this as the tracing can only show that changes occurred - this new application needs to modify the network.

ydjainopensource commented:
Okay no issues we will keep this open

ningbo9 commented:
I have test it commit base on branch master after add some log print, but there is a little strange problem: call `kata-runtime kata-netwowk add-iface` in `netmon` will fail。
`$ sudo docker run --rm --name test -tid alpine sleep 10000
023313cf56b58d3283fefc84f4d1787cac2916e758a1d19012c7ff6a0ad4500e
$ sudo nsenter --net=`docker inspect test |jq -r .[0].NetworkSettings.SandboxKey` ./kata-netmon -d -log info  -r /usr/bin/kata-runtime -s 023313cf56b58d3283fe
fc84f4d1787cac2916e758a1d19012c7ff6a0ad4500e`

I will get error messages after `sudo docker network connect newnetwork test`:
`INFO[0062] 2222 Exec command: &{Path:/usr/bin/kata-runtime Args:[/usr/bin/kata-runtime kata-network add-iface 023313cf56b58d3283fefc84f4d1787cac2916e758a1d190
12c7ff6a0ad4500e /var/run/kata-containers/netmon/sbs/023313cf56b58d3283fefc84f4d1787cac2916e758a1d19012c7ff6a0ad4500e/shared.json] Env:[] Dir: Stdin:<nil> Std
out:<nil> Stderr:<nil> ExtraFiles:[] SysProcAttr:<nil> Process:<nil> ProcessState:<nil> ctx:<nil> lookPathErr:<nil> finished:false childFiles:[] closeAfterSta
rt:[] closeAfterWait:[] goroutine:[] errch:<nil> waitDone:<nil>}  name=kata-netmon pid=22782 sandbox=023313cf56b58d3283fefc84f4d1787cac2916e758a1d19012c7ff6a0
ad4500e source=netmon
FATA[0062] handleEvents()                                error="exit status 1" name=kata-netmon pid=22782 sandbox=023313cf56b58d3283fefc84f4d1787cac2916e758a1
d19012c7ff6a0ad4500e source=netmon`

but it will work if execute the command manually:
`$ sudo nsenter --net=`docker inspect test |jq -r .[0].NetworkSettings.SandboxKey` cat /var/run/kata-containers/netmon/sbs/023313cf56b58d3283fefc84f4d1787cac29
16e758a1d19012c7ff6a0ad4500e/shared.json
{"device":"eth1","name":"eth1","IPAddresses":[{"family":2,"address":"172.18.0.2","mask":"16"}],"mtu":1500,"hwAddr":"02:42:ac:12:00:02"}`

and the new tap `tap1` interface created Successfully
`$ sudo nsenter --net=`docker inspect test |jq -r .[0].NetworkSettings.SandboxKey` ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: ip_vti0@NONE: <NOARP> mtu 1332 qdisc noop state DOWN group default qlen 1
    link/ipip 0.0.0.0 brd 0.0.0.0
9: eth0@if10: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:00:ca:fe:00:00 brd ff:ff:ff:ff:ff:ff link-netnsid 0
11: eth1@if12: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:00:ca:fe:00:01 brd ff:ff:ff:ff:ff:ff link-netnsid 0
47124: tap1@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1500
    link/ether 02:42:ac:12:00:02 brd ff:ff:ff:ff:ff:ff
44374: tap0@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1500
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff`

so maybe there's some permissions issue in netmon or the way of my use is wrong?

sboeuf commented:
@ningbo9 the PR is not completely ready as it does not include the modification inside kata-runtime itself to launch the network monitor.

sboeuf commented:
@ningbo9 @miaoyq
This PR is now ready to be reviewed and tested. I need to add some unit tests in the meantime, but please make sure it works for you.
You'll need to enable the network monitor from configuration.toml after you did `make && sudo make install`.
If you run a container with docker, you should be able to add/remove additional networks to the existing container.
Let me know if you have any questions.

@egernst @bergwolf @jodh-intel @amshinde PTAL

jodh-intel commented on netmon/netmon.go:
```diff
@@ -0,0 +1,613 @@
+	return netlink.RouteSubscribe(n.rtUpdateCh, n.rtDoneCh)
+}
+
+func convertInterface(linkAttrs *netlink.LinkAttrs, addrs []netlink.Addr) Interface {
```
- Could you put a brief comment here explaining how the interface is being converted? Same comment for other functions like `scanNetwork()`, `convertRoutes()`).

- Also, functions like this can be unit-tested I think?

sboeuf replied:
I'll add some comments, and I'll definitely add some unit testing for those.

jodh-intel commented on netmon/netmon.go:
```diff
@@ -0,0 +1,613 @@
+		}
+
+		iface := convertInterface(linkAttrs, addrs)
+		n.netIfaces[linkAttrs.Index] = iface
```
Worth adding a few more log calls for example when an interface is "scanned" / route converted / interface converted?

jodh-intel commented on netmon/netmon.go:
```diff
@@ -0,0 +1,613 @@
+	flag.StringVar(&params.logLevel, "log", "warn",
+		"log messages above specified level: debug, warn, error, fatal or panic")
+
+	flag.Parse()
```
Not sure if it's possible with the `flag` package (?), but it would be useful if running `./kata-netmon -h` could show a sentence or two on what this new component actually does. If flag doesn't allow that, how about switching to https://github.com/urfave/cli (or https://github.com/spf13/cobra :)?

sboeuf replied:
I would like to keep this component simple, and pulling down `urfave/cli` will make it bigger. I'll see if what you ask for is possible with `flag` but I don't want to introduce a heavy package to handle the very simple command line.

jodh-intel replied:
Yeah, I understand. Although it might not be possible for `kata-netmon -h` to show this, we *could* make `kata-netmon help` display the human-readable text, then call `flag.PrintDefaults()`?

sboeuf replied:
I have been able to override `-h`

jodh-intel commented on netmon/netmon.go:
```diff
@@ -0,0 +1,613 @@
+	fields := logrus.Fields{
+		"name":   netmonName,
+		"pid":    os.Getpid(),
+		"source": "netmon",
```
Nice. This should allow it to work well with https://github.com/kata-containers/tests/tree/master/cmd/log-parser.

But thinking about the common code in the other components, what about signal handling? It would be great if this too allowed `SIGUSR1` to dump a backtrace (https://github.com/kata-containers/documentation/blob/master/Developer-Guide.md#troubleshoot-kata-containers).

You can rework `cli/signals.go` like I did for the throttler to avoid code duplication:

- https://github.com/kata-containers/ksm-throttler/pull/41

sboeuf replied:
@jodh-intel in order to factorize the code, we need to define a new package, otherwise I don't see how I can share between the `netmon` and the `cli`. Also, I realized that the PR is pretty big now, and defining a new package and adding this backtrace support can definitely be done in a follow up PR. WDYT?

caoruidong replied:
Agree. Because cli is the main package, we are not able to reuse code in it.

jodh-intel commented on Makefile:
```diff
@@ -225,7 +235,12 @@ define SHOW_ARCH
+netmon: $(NETMON_TARGET_OUTPUT)
+
+$(NETMON_TARGET_OUTPUT): $(SOURCES)
+	$(QUIET_BUILD)(cd $(NETMON_DIR) && go build -i -o $@ -ldflags "-X main.version=$(VERSION)")
```
No install target?

sboeuf replied:
Yes I added it later in the commits, but I fixed it by rebasing and bringing the change directly to this commit.

jodh-intel replied:
Please could you:

- Update `show-summary` so that "binaries to install" also iterates through `$(BINLIBEXECLIST)`.
- Update `show-usage` to include `install-bin-libexec` is listed under "Additional targets".
- Potentially also update the `install` target so that it forces `install-bin-libexec` to be called (`sudo make install` should DTRT I think :)

sboeuf replied:
- I have updated `show-summary`
- I have updated `install` target so that it invokes `install-bin-libexec` also
- I have __not__ updated `show-usage` since any other subcommand such as `install-bin` was already listed. And also, because now `install` takes care of `install-bin-libexec`, I think that's fine.

jodh-intel commented on virtcontainers/sandbox.go:
```diff
@@ -969,6 +969,37 @@ func (s *Sandbox) Delete() error {
 }
 
+func (s *Sandbox) startNetworkMonitor() error {
+	binPath, err := os.Executable()
```
Could you create a trace span here?:

```
func (s *Sandbox) startNetworkMonitor() error {
    span, _ := s.trace("startNetworkMonitor")
    defer span.Finish()

        :
}
```

sboeuf replied:
Yes sure!

jodh-intel commented:
Since this adds a new component, once it lands we'll need updates to atleast the following (but please comment here if you can think of other things we need to do :-)...

- https://github.com/kata-containers/documentation/blob/master/architecture.md
- https://github.com/kata-containers/tests/blob/master/integration/stability/soak_parallel_rm.sh
- https://github.com/kata-containers/runtime/blob/master/data/kata-collect-data.sh.in
- https://github.com/kata-containers/packaging to ensure the new binary is packaged.
- New integration test(s) to exercise the network monitor.

/cc @grahamwhaley, @egernst, @jcvenegas.

sboeuf commented on cli/kata-env_test.go:
```diff
@@ -31,6 +31,7 @@ import (
 const testProxyURL = "file:///proxyURL"
 const testProxyVersion = "proxy version 0.1"
 const testShimVersion = "shim version 0.1"
+const testNetmonVersion = "netmon version 0.1"
```
Sure I can bump version in `kata-env.go`

sboeuf commented on netmon/netmon.go:
```diff
@@ -0,0 +1,613 @@
+// description of the agent protocol. Those definitions need to be in their
+// own separate package so that they can be imported directly from this code.
+// The reason for not importing them now, is because importing the whole agent
+// protocol adds up too much overhead because of the grpc protocol involved.
```
No, the real solution is to move the types out of this protocol package. They need to be in their own package so that we can only import the structure definition. But that's definitely something that needs to be done in parallel, and I didn't want to delay this PR because moving things out might take some time.

jodh-intel replied:
Oh - I see what you mean. Something like this?

```diff
diff --git a/protocols/grpc/agent.proto b/protocols/grpc/agent.proto
index 4c09d9c..f7f27d4 100644
--- a/protocols/grpc/agent.proto
+++ b/protocols/grpc/agent.proto
@@ -10,6 +10,7 @@ package grpc;
 
 import "oci.proto";
 import "google/protobuf/empty.proto";
+import "types.proto";
 
 
 // unstable
@@ -270,42 +271,10 @@ message CreateSandboxRequest {
 message DestroySandboxRequest {
 }
 
-enum IPFamily {
-	v4 = 0;
-	v6 = 1;
-}
-
-message IPAddress {
-	IPFamily family = 1;
-	string address = 2;
-	string mask = 3;
-}
-
-message Interface {
-	string device = 1;
-	string name = 2;
-	repeated IPAddress IPAddresses = 3;
-	uint64 mtu = 4;
-	string hwAddr = 5;
-
-	// pciAddr is the PCI address in the format  "bridgeAddr/deviceAddr".
-	// Here, bridgeAddr is the address at which the bridge is attached on the root bus,
-	// while deviceAddr is the address at which the network device is attached on the bridge.
-	string pciAddr = 6;
-}
-
 message Interfaces {
 	repeated Interface Interfaces = 1;
 }
 
-message Route {
-	string dest = 1;
-	string gateway = 2;
-	string device = 3;
-	string source = 4;
-	uint32 scope = 5;
-}
-
 message Routes {
 	repeated Route Routes = 1;
 }
diff --git a/protocols/grpc/types.proto b/protocols/grpc/types.proto
new file mode 100644
index 0000000..823022f
--- /dev/null
+++ b/protocols/grpc/types.proto
@@ -0,0 +1,35 @@
+syntax = "proto3";
+
+package grpc;
+
+enum IPFamily {
+	v4 = 0;
+	v6 = 1;
+}
+
+message IPAddress {
+	IPFamily family = 1;
+	string address = 2;
+	string mask = 3;
+}
+
+message Interface {
+	string device = 1;
+	string name = 2;
+	repeated IPAddress IPAddresses = 3;
+	uint64 mtu = 4;
+	string hwAddr = 5;
+
+	// pciAddr is the PCI address in the format  "bridgeAddr/deviceAddr".
+	// Here, bridgeAddr is the address at which the bridge is attached on the root bus,
+	// while deviceAddr is the address at which the network device is attached on the bridge.
+	string pciAddr = 6;
+}
+
+message Route {
+	string dest = 1;
+	string gateway = 2;
+	string device = 3;
+	string source = 4;
+	uint32 scope = 5;
+}
```

That seems to work, so why can't we JFDI? :)

jodh-intel replied:
Doh - as @sboeuf has pointed out, although this "works", it doesn't actually do what we want - we want just the types in a `types.go`, we don't want all the gobs of auto-generated functions that grpc puts into `types.pb.go`.

sboeuf commented on netmon/netmon.go:
```diff
@@ -0,0 +1,613 @@
+	netlinkFamily = netlink.FAMILY_V4
+
+	storageParentPath = "/var/run/kata-containers/netmon/sbs"
+	storageDirPerm    = 0750
```
Will do!

sboeuf commented on netmon/netmon.go:
```diff
@@ -0,0 +1,613 @@
+
+	flag.Parse()
+
+	if version {
```
Yes sure.

sboeuf commented on netmon/netmon.go:
```diff
@@ -0,0 +1,613 @@
+
+	for _, netRoute := range netRoutes {
+		dst := ""
+		if netRoute.Dst != nil && netRoute.Dst.IP.To4() != nil {
```
I'll add logs and comments saying that we only support IPv4 for now.

sboeuf commented on netmon/netmon.go:
```diff
@@ -0,0 +1,613 @@
+	// list of interfaces. Otherwise, the deletion will be ignored.
+	linkAttrs := ev.Link.Attrs()
+	if linkAttrs == nil {
+		n.logger().Warnf("Link attributes are nil")
```
I'll change that, thanks!

sboeuf commented on netmon/netmon.go:
```diff
@@ -0,0 +1,613 @@
+func (n *netmon) execKataCmd(subCmd string) error {
+	execCmd := exec.Command(n.runtimePath, kataCmd, subCmd, n.sandboxID, n.sharedFile)
+
+	n.logger().Debugf("Exec command: %+v", execCmd)
```
Ok

sboeuf commented:
@jodh-intel I've addressed all your comments except the handling of SIGUSR1 signal. I'll do this later. And I'll add a bit more unit testing, but this PR is very close from its final version now.
@ningbo9 @miaoyq any chance you can test this and confirm it works for you?

jodh-intel commented:
@sboeuf - ack re signal handling.

jodh-intel commented on cli/kata-env.go:
```diff
@@ -25,7 +25,7 @@ import (
 // XXX: Increment for every change to the output format
 // (meaning any change to the EnvInfo type).
-const formatVersion = "1.0.15"
+const formatVersion = "1.0.16"
```
Thanks - we also need this command to dump the version of the netmon by implementing a function like:

```
func getNetmonInfo(config oci.RuntimeConfig) (NetmonInfo, error)
```

sboeuf commented:
PR updated and ready. (Some follow up to do once merged)

sboeuf commented:
/retest

sboeuf commented:
/recheck

jodh-intel commented:
Thanks @sboeuf.

lgtm

bergwolf commented:
My main concern for the PR is that it introduces two level of binary dependencies. With containerd shimv2, we finally get rid of shim and proxy processes. Now with the net monitor, we are bringing in `kata-netmon` and `kata-runtime` dependency even for containerd shimv2. What I would really like to see is the ability for the net monitor process not to depend on `kata-runtime` binary but that can be done in future PRs.

Another question is, is there a matching test PR so that we verify this in CI? I'm having issues in testing network hotplug with vm factory. So I'm wondering if network hotplug is working for netns monitoring.

jodh-intel commented:
@bergwolf - +1 for more tests. I've raised https://github.com/kata-containers/tests/issues/728 for that.

sboeuf commented:
@bergwolf 
>My main concern for the PR is that it introduces two level of binary dependencies. With containerd shimv2, we finally get rid of shim and proxy processes. Now with the net monitor, we are bringing in kata-netmon and kata-runtime dependency even for containerd shimv2. What I would really like to see is the ability for the net monitor process not to depend on kata-runtime binary but that can be done in future PRs.

Yes I understand, but the containerd-shim-v2 implementation is not there yet, and `kata-runtime` is still the default way of using Kata as an OCI runtime, so I think that for now at least, the network monitor will solve the `docker connect` missing feature for people interested in having this available.
But don't miss my point here, I think that containerd-shim-v2 will be a better approach since we won't need a separate binary, and we'll be able to handle this directly from a go routine instead ;)

sboeuf commented:
@amshinde @bergwolf @caoruidong @WeiZhang555 PTAL
@ningbo9 @miaoyq sorry to ping you again on this, but it'd be great if you could test this out as you might use it soon.

caoruidong commented on virtcontainers/network.go:
```diff
@@ -143,6 +143,8 @@ type NetworkInterfacePair struct {
 	NetNSPath         string
 	NetNsCreated      bool
+	UseNetmon         bool
+	NetmonConfig      NetmonConfig
```
Just a small question. In some structs, the bool field "enable" is in the config struct. Why they are separate here and in oci/utils.go?

sboeuf replied:
Just made the modification as I agree it didn't make any sense to separate this field from the `NetmonConfig` structure.

bergwolf commented on virtcontainers/sandbox.go:
```diff
@@ -1056,6 +1102,7 @@ func (s *Sandbox) AddInterface(inf *grpc.Interface) (*grpc.Interface, error) {
 	}
 
 	// Add network for vm
+	inf.PciAddr = endpoint.PCIAddr
```
AFAICT, this is not enough. The endpoint pci address is determined when qemu plugs it to the guest vm but we are not saving it back to the endpoint data structure. You need something like https://github.com/bergwolf/kata-runtime/commit/2e3fb6f34963b1afe886b30dd6dfaaa2791d6e08 to make it work.

sboeuf replied:
@bergwolf https://github.com/bergwolf/kata-runtime/commit/2e3fb6f34963b1afe886b30dd6dfaaa2791d6e08 looks a lot like https://github.com/kata-containers/runtime/pull/534/commits/00b48c9c5d77843fb8c5a75a1ed5cc3fed5a8728, where's the diff.
And yes, I have tested this whole PR, I'm able to add and remove additional network interfaces with `docker connect ...`.

bergwolf replied:
Ok, I missed that. Sorry for the noise.

bergwolf commented:
@sboeuf I tried the PR locally with `docker network connect...` and hit agent crash:
> Sep 13 09:21:01 macbeth kata-proxy[53939]: time="2018-09-13T09:21:01.01754966+08:00" level=info msg="time=\"2018-09-13T01:21:00.997030557Z\" level=debug msg=\"new request\" name=kata-agent pid=128 req=\"interface:<device:\\\"eth1\\\" name:\\\"eth1\\\" IPAddresses:<family:2 address:\\\"172.18.0.2\\\" mask:\\\"16\\\" > mtu:1500 hwAddr:\\\"02:42:ac:12:00:02\\\" pciAddr:\\\"02/01\\\" > \" request=/grpc.AgentService/UpdateInterface sandbox=a6ec2b4095158602d344e07f0bb974f1a42c5dda2f63cf1901fbe4fb30072e5b source=agent\n" name=kata-proxy pid=53939 sandbox=a6ec2b4095158602d344e07f0bb974f1a42c5dda2f63cf1901fbe4fb30072e5b source=agent
Sep 13 09:21:01 macbeth kata-proxy[53939]: time="2018-09-13T09:21:01.018416928+08:00" level=info msg="time=\"2018-09-13T01:21:00.997952292Z\" level=info msg=\"Fetched PCI address for device\" completePCIAddr=\"0000:00:02.0/0000:01:01.0\" name=kata-agent pid=128 sandbox=a6ec2b4095158602d344e07f0bb974f1a42c5dda2f63cf1901fbe4fb30072e5b source=agent\n" name=kata-proxy pid=53939 sandbox=a6ec2b4095158602d344e07f0bb974f1a42c5dda2f63cf1901fbe4fb30072e5b source=agent
Sep 13 09:21:01 macbeth kata-proxy[53939]: time="2018-09-13T09:21:01.019656886+08:00" level=info msg="time=\"2018-09-13T01:21:00.998785298Z\" level=info msg=\"Waiting on channel for device notification\" name=kata-agent pciID=02/01 pid=128 sandbox=a6ec2b4095158602d344e07f0bb974f1a42c5dda2f63cf1901fbe4fb30072e5b source=agent\n" name=kata-proxy pid=53939 sandbox=a6ec2b4095158602d344e07f0bb974f1a42c5dda2f63cf1901fbe4fb30072e5b source=agent
Sep 13 09:21:04 macbeth kata-proxy[53939]: time="2018-09-13T09:21:04.02648783+08:00" level=info msg="panic: close of nil channel\n" name=kata-proxy pid=53939 sandbox=a6ec2b4095158602d344e07f0bb974f1a42c5dda2f63cf1901fbe4fb30072e5b source=agent
Sep 13 09:21:04 macbeth kata-proxy[53939]: time="2018-09-13T09:21:04.027799889+08:00" level=info msg="\n" name=kata-proxy pid=53939 sandbox=a6ec2b4095158602d344e07f0bb974f1a42c5dda2f63cf1901fbe4fb30072e5b source=agent
Sep 13 09:21:04 macbeth kata-proxy[53939]: time="2018-09-13T09:21:04.029840616+08:00" level=info msg="goroutine 78 [running]:\n" name=kata-proxy pid=53939 sandbox=a6ec2b4095158602d344e07f0bb974f1a42c5dda2f63cf1901fbe4fb30072e5b source=agent
Sep 13 09:21:04 macbeth kata-proxy[53939]: time="2018-09-13T09:21:04.034235867+08:00" level=info msg="main.getPCIDeviceName(0xc42008f1e0, 0xc42044d1e0, 0x5, 0xb, 0xc42014ebe8, 0x0, 0x2b8)\n" name=kata-proxy pid=53939 sandbox=a6ec2b4095158602d344e07f0bb974f1a42c5dda2f63cf1901fbe4fb30072e5b source=agent

And even after I fixed the crash and a few other issues, see kata-containers/agent#367, I still get network hotplug timeout errors:
> Sep 13 09:37:20 macbeth kata-proxy[54869]: time="2018-09-13T09:37:20.532400733+08:00" level=info msg="time=\"2018-09-13T01:37:20.511006103Z\" level=debug msg=\"new request\" name=kata-agent pid=128 req=\"interface:<device:\\\"eth1\\\" name:\\\"eth1\\\" IPAddresses:<family:2 address:\\\"172.18.0.2\\\" mask:\\\"16\\\" > mtu:1500 hwAddr:\\\"02:42:ac:12:00:02\\\" pciAddr:\\\"02/01\\\" > \" request=/grpc.AgentService/UpdateInterface sandbox=e2145826daa97ea538d49bffc8c25ad13e0849a1edf5546478458fa17aa63002 source=agent\n" name=kata-proxy pid=54869 sandbox=e2145826daa97ea538d49bffc8c25ad13e0849a1edf5546478458fa17aa63002 source=agent
Sep 13 09:37:20 macbeth kata-proxy[54869]: time="2018-09-13T09:37:20.533607734+08:00" level=info msg="time=\"2018-09-13T01:37:20.512336531Z\" level=info msg=\"Fetched PCI address for device\" completePCIAddr=\"0000:00:02.0/0000:01:01.0\" name=kata-agent pid=128 sandbox=e2145826daa97ea538d49bffc8c25ad13e0849a1edf5546478458fa17aa63002 source=agent\n" name=kata-proxy pid=54869 sandbox=e2145826daa97ea538d49bffc8c25ad13e0849a1edf5546478458fa17aa63002 source=agent
Sep 13 09:37:20 macbeth kata-proxy[54869]: time="2018-09-13T09:37:20.535074125+08:00" level=info msg="time=\"2018-09-13T01:37:20.513595652Z\" level=info msg=\"Waiting on channel for device notification\" name=kata-agent pciID=02/01 pid=128 sandbox=e2145826daa97ea538d49bffc8c25ad13e0849a1edf5546478458fa17aa63002 source=agent\n" name=kata-proxy pid=54869 sandbox=e2145826daa97ea538d49bffc8c25ad13e0849a1edf5546478458fa17aa63002 source=agent
Sep 13 09:37:23 macbeth kata-proxy[54869]: time="2018-09-13T09:37:23.537970764+08:00" level=info msg="time=\"2018-09-13T01:37:23.51786459Z\" level=debug msg=\"request end\" duration=3.005565104s error=\"rpc error: code = DeadlineExceeded desc = Timeout reached after 3s waiting for device 0000:00:02.0/0000:01:01.0\" name=kata-agent pid=128 request=/grpc.AgentService/UpdateInterface resp=\"<nil>\" sandbox=e2145826daa97ea538d49bffc8c25ad13e0849a1edf5546478458fa17aa63002 source=agent\n" name=kata-proxy pid=54869 sandbox=e2145826daa97ea538d49bffc8c25ad13e0849a1edf5546478458fa17aa63002 source=agent
Sep 13 09:37:23 macbeth kata-runtime[54984]: time="2018-09-13T09:37:23.540837194+08:00" level=error msg="update interface request failed" arch=amd64 command=kata-network container=e2145826daa97ea538d49bffc8c25ad13e0849a1edf5546478458fa17aa63002 error="rpc error: code = DeadlineExceeded desc = Timeout reached after 3s waiting for device 0000:00:02.0/0000:01:01.0" interface-requested="device:\"eth1\" name:\"eth1\" IPAddresses:<family:2 address:\"172.18.0.2\" mask:\"16\" > mtu:1500 hwAddr:\"02:42:ac:12:00:02\" pciAddr:\"02/01\" " name=kata-runtime pid=54984 resulting-interface="<nil>" sandbox=e2145826daa97ea538d49bffc8c25ad13e0849a1edf5546478458fa17aa63002 source=virtcontainers subsystem=kata_agent
Sep 13 09:37:23 macbeth kata-runtime[54984]: time="2018-09-13T09:37:23.540955376+08:00" level=error msg="add interface failed" arch=amd64 command=kata-network container=e2145826daa97ea538d49bffc8c25ad13e0849a1edf5546478458fa17aa63002 error="rpc error: code = DeadlineExceeded desc = Timeout reached after 3s waiting for device 0000:00:02.0/0000:01:01.0" name=kata-runtime pid=54984 resulting-interface="<nil>" sandbox=e2145826daa97ea538d49bffc8c25ad13e0849a1edf5546478458fa17aa63002 source=runtime
Sep 13 09:37:23 macbeth kata-runtime[54984]: time="2018-09-13T09:37:23.541177053+08:00" level=error msg="rpc error: code = DeadlineExceeded desc = Timeout reached after 3s waiting for device 0000:00:02.0/0000:01:01.0" arch=amd64 command=kata-network container=e2145826daa97ea538d49bffc8c25ad13e0849a1edf5546478458fa17aa63002 name=kata-runtime pid=54984 sandbox=e2145826daa97ea538d49bffc8c25ad13e0849a1edf5546478458fa17aa63002 source=runtime

For complete runtime/agent log please see: https://pastebin.com/JV5Prj7W

Could you tell how you tested the PR?

sboeuf commented:
@bergwolf ok so I am gonna tell you how I am testing the PR, I actually just did right now in order to validate what I'm about to say. But as I told you, I am not saying the uevent handler is not foolproof, and we might need a few patches in there.

- So first get this PR and build it (`make`)
- Install binaries (`sudo make install`). You should have `kata-netmon` installed inside `/usr/libexec/kata-containers/kata-netmon`
- Make sure your `configuration.toml` includes:
```toml
[netmon]
# If enabled, the network monitoring process gets started when the
# sandbox is created. This allows for the detection of some additional
# network being added to the existing network namespace, after the
# sandbox has been created.
# (default: disabled)
enable = true

# Specify the path to the netmon binary.
path = "/usr/libexec/kata-containers/kata-netmon"

# If enabled, netmon messages will be sent to the system log
# (default: disabled)
#enable_debug = true
```
- Then creates a docker interface (I picked up a bridge)
- Start your docker container `docker run -it --rm --runtime=kata-runtime busybox` and check the interfaces available:
```
/ # ifconfig
eth0      Link encap:Ethernet  HWaddr 02:42:AC:11:00:02  
          inet addr:172.17.0.2  Bcast:172.17.255.255  Mask:255.255.0.0
          inet6 addr: fe80::42:acff:fe11:2/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:8 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:656 (656.0 B)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
```
- From a different shell, add the interface to the container:
```
$ docker network connect br0 $(docker ps -a -q)
```
- Check the network again from the container:
```
/ # ifconfig
eth0      Link encap:Ethernet  HWaddr 02:42:AC:11:00:02  
          inet addr:172.17.0.2  Bcast:172.17.255.255  Mask:255.255.0.0
          inet6 addr: fe80::42:acff:fe11:2/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:7 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:586 (586.0 B)

eth1      Link encap:Ethernet  HWaddr 02:42:C1:A9:00:02  
          inet addr:193.169.0.2  Bcast:193.169.255.255  Mask:255.255.0.0
          inet6 addr: fe80::42:c1ff:fea9:2/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:2 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:180 (180.0 B)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

```
- You should be able to ping the bridge IP address from inside the container.

bergwolf commented:
@sboeuf Thanks. That's exactly what I did locally. The only difference is that I created a new bridge network with `docker network create foobarnetwork`:
```
[macbeth@qapi]$docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
9be9da61588f        bridge              bridge              local
d2e4c7b0b1a8        foobarnetwork       bridge              local
51e2315a5c83        host                host                local
390b1f61c043        none                null                local
```
Then connect to it with `docker network connect foobarnetwork e2145826daa9`. The connect command did return immediately but the network is not brought up and from the logs I saw the failure.

And surely I have netmon enabled in the config:
```
[macbeth@runtime]$cat /etc/kata-containers/configuration.toml |grep netmon -A 10
[netmon]
# If enabled, the network monitoring process gets started when the
# sandbox is created. This allows for the detection of some additional
# network being added to the existing network namespace, after the
# sandbox has been created.
# (default: disabled)
enable = true

# Specify the path to the netmon binary.
path = "/usr/libexec/kata-containers/kata-netmon"

# If enabled, netmon messages will be sent to the system log
# (default: disabled)
#enable_debug = true
```
And I'm using image rootfs rather than initrd since I assume you are not testing with initrd.
```
[macbeth@virtcontainers]$cat /etc/kata-containers/configuration.toml |grep image
image = "/golang/src/github.com/kata-containers/osbuilder/kata-containers.img"
```

Not sure why I see these errors. Could you please check the log to see if the new interface is brought up by `UpdateInterface`? Are you running the latest agent master? What rootfs image distro are you using?

I tried it on another test machine and got the same errors. I wonder what difference makes it always fail on my side. TBF I've been chasing the failure for some time and it's great to know that it works on your side.

btw, here is my `kata-env` result:
```
[vmfarm@runtime]$kata-runtime kata-env
[Meta]
  Version = "1.0.16"

[Runtime]
  Debug = false
  Path = "/home/bergwolf/workplace/golang/src/github.com/kata-containers/runtime/kata-runtime"
  [Runtime.Version]
    Semver = "1.3.0-rc0"
    Commit = "37d9971c3fb216e43333d4d4ad01870f0544848b"
    OCI = "1.0.1"
  [Runtime.Config]
    Path = "/etc/kata-containers/configuration.toml"

[Hypervisor]
  MachineType = "q35"
  Version = "QEMU emulator version 2.11.0 (v2.11.0-28-g6bfb0a7-dirty)\nCopyright (c) 2003-2017 Fabrice Bellard and the QEMU Project developers"
  Path = "/home/bergwolf/src/qemu/x86_64-softmmu/qemu-system-x86_64"
  BlockDeviceDriver = "virtio-scsi"
  Msize9p = 1048576
  Debug = false
  UseVSock = false

[Image]
  Path = "/home/bergwolf/workplace/golang/src/github.com/kata-containers/osbuilder/kata-containers.img"

[Kernel]
  Path = "/usr/lib/kata-containers/vmlinuz-4.14.29"
  Parameters = "vsyscall=emulate agent.log=debug"

[Initrd]
  Path = ""

[Proxy]
  Type = "kataProxy"
  Version = "kata-proxy version 1.3.0-rc1-981fef4774ba15cf94a3a9013629d0ab60668348"
  Path = "/usr/libexec/kata-containers/kata-proxy"
  Debug = true

[Shim]
  Type = "kataShim"
  Version = "kata-shim version 1.3.0-rc1-9b2891cfb153967fa4a65e44b2928255c889f643"
  Path = "/home/bergwolf/workplace/golang/src/github.com/kata-containers/shim/kata-shim"
  Debug = true

[Agent]
  Type = "kata"

[Host]
  Kernel = "4.9.13"
  Architecture = "amd64"
  VMContainerCapable = true
  SupportVSocks = false
  [Host.Distro]
    Name = "Ubuntu"
    Version = "16.04"
  [Host.CPU]
    Vendor = "GenuineIntel"
    Model = "Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz"

[Netmon]
  Version = "kata-netmon version 1.3.0-rc0"
  Path = "/usr/libexec/kata-containers/kata-netmon"
  Debug = true
  Enable = true
```

bergwolf commented on cli/config/configuration.toml.in:
```diff
@@ -177,6 +177,21 @@ path = "@SHIMPATH@"
+# network being added to the existing network namespace, after the
+# sandbox has been created.
+# (default: disabled)
+#enable = true
```
nits: as I put a test commit at https://github.com/kata-containers/tests/pull/743, I noticed that it would be better to call this `enable_netmon` instead, so that we can easily grep and replace it in scripts.

sboeuf replied:
Of course I can do that and I agree this might help!

bergwolf commented:
btw, @sboeuf if you want to merge this PR feel free to do so. It would be great if you can rename the `enable` option to `enable_netmon` though.

Anyway, whatever error I'm seeing w/ network hotplug, it is not related to your PR. So I'm giving my
LGTM

bergwolf commented:
@caoruidong I've tried alpine based initrd w/ agent as init, and centos based rootfs image w/ systemd as init, both failed the same way. It seems https://github.com/kata-containers/tests/pull/743 have given the same results:
```
+ sudo docker network create foobartestnetwork
4485b0308d12816eabc6ffbb900228c12b349dee3866aa1211ec1d946c5c0511
+ sudo docker run --runtime=kata-runtime -d --name kata-vKBoFN busybox tail -f /dev/null
b3c68f2d9c37de4176f1de711cc43b090632c744d95f17196643c0fc983dac33
+ sudo docker network connect foobartestnetwork kata-vKBoFN
+ sleep 10
++ sudo docker exec -it kata-vKBoFN ip addr show eth1
++ sed -ne 's/.*inet \([.0-9]\{7,15\}\)\/[0-9]\{1,2\} .*/\1/p'
the input device is not a TTY
+ ipaddr=
+ sudo docker rm -f kata-vKBoFN
kata-vKBoFN
+ sudo docker network rm foobartestnetwork
foobartestnetwork
+ echo hotplugged network ip address is
hotplugged network ip address is
+ test -n ''
```

@sboeuf what guest rootfs distro are you testing with?

caoruidong commented on netmon/netmon.go:
```diff
@@ -0,0 +1,661 @@
+type Interface struct {
+	Device      string       `json:"device,omitempty"`
+	Name        string       `json:"name,omitempty"`
+	IPAddresses []*IPAddress `json:"IPAddresses,omitempty"`
```
Capitalization? OK.

sboeuf replied:
Oh! You mean from the json definition?
Yes I'll change that!

caoruidong replied:
Yes. Thanks.

sboeuf replied:
Oh wait, I did that so that it comply with the structure defined in `agent/protocols/grpc/`. I'm afraid the unmarshalling might not work if I change this.

sboeuf commented:
@amshinde PTAL and feel free to merge if you're fine with it.

amshinde commented on netmon/netmon.go:
```diff
@@ -0,0 +1,565 @@
+
+	// First, ignore if the interface name contains "kata". This way we
+	// are preventing from adding interfaces created by Kata Containers.
+	if strings.Contains(linkAttrs.Name, kataSuffix) {
```
Why not just use strings.HasSuffix here?

sboeuf replied:
done

sboeuf commented on netmon/netmon.go:
```diff
@@ -0,0 +1,565 @@
+	}
+
+	// First, ignore if the interface name contains "kata". This way we
+	// are preventing from adding interfaces created by Kata Containers.
```
done

sboeuf commented on netmon/netmon.go:
```diff
@@ -0,0 +1,565 @@
+	// For simplicity the code will only focus on IPv4 addresses for now.
+	netlinkFamily = netlink.FAMILY_V4
+
+	storageParentPath = "/var/run/kata-containers/netmon/sbs"
```
The Makefile does not contain any other storage path yet. Should this be part of a more global PR pushing the default storage path to the Makefile?
WDYT?

sboeuf commented:
@amshinde changes applied, PTAL.

amshinde commented:
lgtm.
@sboeuf CI seems to be failing on ubuntu.

sboeuf commented:
@amshinde yeah I've just restarted because three builds failed because of networking issues. Nothing related to the PR.

sboeuf commented:
@amshinde could you please merge this one?
